smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/ServiceClientConfigGenerator.kt

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package software.amazon.smithy.kotlin.codegen.rendering
+
+import software.amazon.smithy.codegen.core.CodegenException
+import software.amazon.smithy.kotlin.codegen.core.CodegenContext
+import software.amazon.smithy.kotlin.codegen.core.KotlinWriter
+import software.amazon.smithy.kotlin.codegen.model.expectTrait
+import software.amazon.smithy.kotlin.codegen.model.hasIdempotentTokenMember
+import software.amazon.smithy.kotlin.codegen.model.hasTrait
+import software.amazon.smithy.kotlin.codegen.rendering.endpoints.DefaultEndpointProviderGenerator
+import software.amazon.smithy.kotlin.codegen.rendering.endpoints.EndpointProviderGenerator
+import software.amazon.smithy.kotlin.codegen.rendering.util.AbstractConfigGenerator
+import software.amazon.smithy.kotlin.codegen.rendering.util.ConfigProperty
+import software.amazon.smithy.kotlin.codegen.rendering.util.ConfigPropertyType
+import software.amazon.smithy.kotlin.codegen.rendering.util.RuntimeConfigProperty
+import software.amazon.smithy.kotlin.codegen.utils.getOrNull
+import software.amazon.smithy.kotlin.codegen.utils.toCamelCase
+import software.amazon.smithy.model.shapes.ServiceShape
+import software.amazon.smithy.model.shapes.ShapeType
+import software.amazon.smithy.rulesengine.traits.ClientContextParamsTrait
+import software.amazon.smithy.rulesengine.traits.EndpointRuleSetTrait
+
+/**
+ * Default generator for rendering a service config. By default integrations can register additional properties
+ * without overriding the entirety of service config generation.
+ *
+ * @param serviceShape the service shape to render config for
+ * @param detectDefaultProps Flag indicating if properties should be added automatically based on model detection
+ */
+class ServiceClientConfigGenerator(
+    private val serviceShape: ServiceShape,
+    private val detectDefaultProps: Boolean = true,
+) : AbstractConfigGenerator() {
+
+    /**
+     * Attempt to detect configuration properties automatically based on the model
+     */
+    private fun detectDefaultProps(context: CodegenContext, shape: ServiceShape): List<ConfigProperty> = buildList {
+        add(RuntimeConfigProperty.SdkLogMode)
+        if (context.protocolGenerator?.applicationProtocol?.isHttpProtocol == true) {
+            add(RuntimeConfigProperty.HttpClientEngine)
+            add(RuntimeConfigProperty.HttpInterceptors)
+        }
```

Comment:
**Question**: I wonder if this belongs in a subclass or the protocol generator itself. In theory, every protocol may have its own set of configuration parameters. Did you consider any other locations to add protocol-specific config?

Reply:
I did not, this was existing code that was moved around. It could theoretically be moved around at some point though.